### PR TITLE
 plan: fix a bug when expectedCnt and count after filter are all 0.

### PR DESCRIPTION
**Pull request by winoros**
In this case, we cannot set the first read's row count to 0 because it reads data in fact.
PTAL @lamxTyler @shenli @zz-jason

### CODE REVIEW

**Comment by shenli** [comment]
Please add a test case.

**Comment by zz-jason** [comment]
@winoros ci failed.

File: plan/physical_plan_builder.go [hunk 1]
```diff
@@ -354,7 +354,7 @@ func (ds *DataSource) convertToIndexScan(prop *requiredProp, idx *model.IndexInf
 			}
 		}
 	}
-	if matchProperty && prop.expectedCnt < math.MaxFloat64 {
+	if matchProperty && prop.expectedCnt < math.MaxFloat64 && prop.expectedCnt > ds.statsAfterSelect.count {
```
Comment (lamxTyler) [inline plan/physical_plan_builder.go]: I think that it is fine for `expectCnt` not larger than `statsAfterSelect.count`.

**Comment by winoros** [comment]
@shenli added
@lamxTyler fixed

**Comment by zz-jason** [comment]
LGTM

File: plan/physical_plan_builder.go [hunk 2]
```diff
@@ -354,7 +354,7 @@ func (ds *DataSource) convertToIndexScan(prop *requiredProp, idx *model.IndexInf
 			}
 		}
 	}
-	if matchProperty && prop.expectedCnt < math.MaxFloat64 {
+	if matchProperty && prop.expectedCnt < ds.statsAfterSelect.count {
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Add a comment for this condition.

File: plan/physical_plan_builder.go [hunk 3]
```diff
@@ -572,7 +572,7 @@ func (ds *DataSource) convertToTableScan(prop *requiredProp) (task task, err err
 	}
 	task = copTask
 	matchProperty := len(prop.cols) == 1 && pkCol != nil && prop.cols[0].Equal(nil, pkCol)
-	if matchProperty && prop.expectedCnt < math.MaxFloat64 {
+	if matchProperty && prop.expectedCnt < ds.statsAfterSelect.count {
```
Comment (shenli) [inline plan/physical_plan_builder.go]: Ditto.

**Comment by winoros** [comment]
@shenli added.

**Comment by zz-jason** [comment]
@lamxTyler PTAL

File: plan/physical_plan_builder.go [hunk 4]
```diff
@@ -354,7 +354,10 @@ func (ds *DataSource) convertToIndexScan(prop *requiredProp, idx *model.IndexInf
 			}
 		}
 	}
-	if matchProperty && prop.expectedCnt < math.MaxFloat64 {
+	// Only use expectedCnt when it's smaller than the count we calculated.
+	// e.g. IndexScan(count1)->After Filter(count2). The `ds.statsAfterSelect.count` is count2. count1 is the one we need to calculate
+	// If expectedCnt and count2 are both zero and we go into the if block, the count1 will be set to zero though it's shouldn't be.
```
Comment (zimulala) [inline plan/physical_plan_builder.go]: The `if` is hard to understand.

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
@winoros Why expectedCount is 0 when `LIMIT` is 1?

**Comment by winoros** [comment]
@coocood The limit is beyond the aggregation. After the aggregation, the expectedCnt become 0.

**Comment by winoros** [comment]
@coocood  PTAL

**Pull request merged**